### ISSUE TITLE
Campaign Emails not being sent

### ISSUE THREAD

**Issue by AmitBFG** [mautic/mautic#5546]
**Please DO NOT report security vulnerabilities here. Send them to security@mautic.com instead.**

What type of report is this:

| Q  | A
| ---| ---
| Bug report? | Yes
| Feature request? | No
| Enhancement? | No

## Description:

Hi,

I have setup Mautic on my server. Configured all cron jobs.

Email also works when somone fills form on my website. The problem im facing is, emails are not being sent through campaigns. I have tried to troubleshoot for last few days, reading and re-reading for a resolution.

I believe my cron jobs are accurately configured. Im pasting it here in anycase.

*/15	*	*	*	*	/usr/local/cpanel/3rdparty/bin/php /home/simplif6/public_html/yfret/app/console mautic:broadcasts:send > /dev/null 2>&1

3,18,33,48	*	*	*	*	/usr/local/cpanel/3rdparty/bin/php /home/simplif6/public_html/yfret/app/console mautic:webhooks:process > /dev/null 2>&1

0,15,30,45	*	*	*	*	/usr/local/cpanel/3rdparty/bin/php /home/simplif6/public_html/yfret/app/console mautic:segments:update > /dev/null 2>&1

5,20,35,50	*	*	*	*	/usr/local/cpanel/3rdparty/bin/php /home/simplif6/public_html/yfret/app/console mautic:campaigns:update > /dev/null 2>&1

10,25,40,55	*	*	*	*	/usr/local/cpanel/3rdparty/bin/php /home/simplif6/public_html/yfret/app/console mautic:campaigns:trigger > /dev/null 2>&1

5,20,35,50	*	*	*	*	/usr/local/cpanel/3rdparty/bin/php /home/simplif6/public_html/yfret/app/console mautic:campaigns:rebuild > /dev/null 2>&1

*/15	*	*	*	*	/usr/local/cpanel/3rdparty/bin/php /home/simplif6/public_html/yfret/app/console mautic:emails:send > /dev/null 2>&1

*/15	*	*	*	*	/usr/local/cpanel/3rdparty/bin/php /home/simplif6/public_html/yfret/app/console mautic:messages:send > /dev/null 2>&1

The campaign triggers the email, as i can see it in Mautic calendar, but the actual email doesn’t go through.

Im not an expert at understanding/ debugging based on logs. Can anyone pls help me resolve this.

Thanks for your help.
## If a bug:

| Q   | A
| --- | ---
| Mautic version | 2.12.0
| PHP version | 7.0

### Steps to reproduce:
1. Create a campaign
2. Select a campaign segment as contact source
3. Send email at defined time or interval or immediately.
 
### Log errors: 

_Please check for related errors in the latest log file in [mautic root]/app/log/ and/or the web server's logs and post them here. Be sure to remove sensitive information if applicable._

**Comment by escopecz** [comment]
1. What email provider do you use?
2. Can you switch to another one just to rule out it's problem in that provider?
3. Did you configure the email queue or are you sending immediately?
4. Are there any errors in the logs?

**Comment by void-gh** [comment]
We're experiencing the same issue with Campaign emails not being sent. As with the original poster the event is triggered, and the cron logs show the queue being processed; however, no email is actually delivered. Mautic version 2.12.1 and PHP is 7.1.11. 

1. We're using our own email server.
2. Test email's work fine from the Configuration > Email Settings page, and messages sent out to segments immediately work fine.
3. Mail is configured for Queue.
4. We have the following errors in the apps/logs/mautic_prod-YYYY-MM-DD log:

```
[2018-02-20 16:00:01] mautic.NOTICE: Swift_TransportException: Expected response code 220 but got code "421", with message "421 4.3.2 Service not available " (uncaught exception) at /var/www/mautic/vendor/swiftmailer/swiftmailer/lib/classes/Swift/Transport/AbstractSmtpTransport.php line 383 while running console command `mautic:emails:send` [] []
[2018-02-20 16:00:01] mautic.WARNING: Command `mautic:emails:send` exited with status code 255 [] []
[2018-02-20 16:30:02] mautic.NOTICE: Swift_TransportException: Expected response code 220 but got code "421", with message "421 4.3.2 Service not available " (uncaught exception) at /var/www/mautic/vendor/swiftmailer/swiftmailer/lib/classes/Swift/Transport/AbstractSmtpTransport.php line 383 while running console command `mautic:emails:send` [] []
[2018-02-20 16:30:02] mautic.WARNING: Command `mautic:emails:send` exited with status code 255 [] []
[2018-02-20 17:00:02] mautic.NOTICE: Swift_IoException: Unable to open file for reading [blob:https://fqdn.of.host.com/e1fd1ded-09e8-403b-9aae-7fc9e9f5a8ed] (uncaught exception) at /var/www/mautic/vendor/swiftmailer/swiftmailer/lib/classes/Swift/ByteStream/FileByteStream.php line 144 while running console command `mautic:emails:send` [] []
```

And our cron jobs are configured as follows:

```
#### Keep segments current
0,15,30,45     *    *   *   *   /usr/bin/php /var/www/mautic/app/console mautic:segments:update --no-interaction --no-ansi >> /var/log/mautic/mautic-cron.log
#### Keep campaigns updated with applicable contacts
5,20,35,50     *    *   *   *   /usr/bin/php /var/www/mautic/app/console mautic:campaigns:rebuild --no-interaction --no-ansi >> /var/log/mautic/mautic-cron.log
8,23,38,53     *    *   *   *   /usr/bin/php /var/www/mautic/app/console mautic:campaigns:update --no-interaction --no-ansi >> /var/log/mautic/mautic-cron.log
#### Execute Campaign events
10,25,40,55    *    *   *   *   /usr/bin/php /var/www/mautic/app/console mautic:campaigns:trigger --no-interaction --no-ansi >> /var/log/mautic/mautic-cron.log
#### Send Scheduled Broadcasts (e.g. segment emails)
15,30,45,0     *    *   *   *   /usr/bin/php /var/www/mautic/app/console mautic:broadcasts:send --no-interaction --no-ansi >> /var/log/mautic/mautic-cron.log
#### Process Marketing Message queue
0,30           *    *   *   *   /usr/bin/php /var/www/mautic/app/console mautic:messages:send --force --no-interaction --no-ansi >> /var/log/mautic/mautic-cron.log
#### Process Email queue
0,30           *    *   *   *   /usr/bin/php /var/www/mautic/app/console mautic:emails:send --no-interaction --no-ansi >> /var/log/mautic/mautic-cron.log
#### Fetch and Process Monitored Email
0,30           *    *   *   *   /usr/bin/php /var/www/mautic/app/console mautic:email:fetch --no-interaction --no-ansi >> /var/log/mautic/mautic-cron.log
```


And configuring a test campaign just to show the cron logs...

```
[2018-02-20 17:30:02] 0 emails processed for {fqdn.to.mail.server.com:143/imap/tls}INBOX with a criteria of UNSEEN
[2018-02-20 17:35:01] Rebuilding contacts for campaign 7
[2018-02-20 17:35:01] 3 total contact(s) to be added in batches of 300
 0/3 [>---------------------------]   0%
 1/3 [=========>------------------]  33%
 2/3 [==================>---------]  66%
 3/3 [============================] 100%
[2018-02-20 17:35:02] 0 total contact(s) to be removed in batches of 300
[2018-02-20 17:35:02] 3 contact(s) affected

[2018-02-20 17:40:01] Triggering events for campaign 7
[2018-02-20 17:40:01] Triggering first level events
[2018-02-20 17:40:01] 3 total events(s) to be processed in batches of 100
 0/3 [>---------------------------]   0%
 1/3 [=========>------------------]  33%
 2/3 [==================>---------]  66%
 3/3 [============================] 100%
[2018-02-20 17:40:01] 3 event(s) executed

[2018-02-20 17:50:01] Rebuilding contacts for campaign 7
[2018-02-20 17:50:01] 0 total contact(s) to be added in batches of 300
[2018-02-20 17:50:01] 0 total contact(s) to be removed in batches of 300
[2018-02-20 17:50:01] 0 contact(s) affected

[2018-02-20 17:53:02] Rebuilding contacts for campaign 7
[2018-02-20 17:53:02] 0 total contact(s) to be added in batches of 300
[2018-02-20 17:53:02] 0 total contact(s) to be removed in batches of 300
[2018-02-20 17:53:02] 0 contact(s) affected

[2018-02-20 17:55:01] Triggering events for campaign 7
[2018-02-20 17:55:01] Triggering first level events
[2018-02-20 17:55:01] 0 total events(s) to be processed in batches of 100
[2018-02-20 17:55:01] 0 event(s) executed

[2018-02-20 17:55:01] Triggering scheduled events
[2018-02-20 17:55:01] 3 total events(s) to be processed in batches of 100
 0/3 [>---------------------------]   0%
 1/3 [=========>------------------]  33%
 2/3 [==================>---------]  66%
 3/3 [============================] 100%
[2018-02-20 17:55:01] 2 event(s) executed
```


And of course, the timestamps reported in the logs don't actually match the server's timezone, nor the "Default timezone" configured within the Mautic UI. If anyone has the answer on how to get the logs to report in the local timezone vs UTC I'd greatly appreciate it.

**Comment by pcottrell** [comment]
I am experiencing a wierd issue where **segment emails** are not sending for certain contacts.

I will send a segment email and watch 4/5 emails send and 1 stays in pending, and I never see it spool.

I am not sure if it is related. I have not looked into whether or not the campaigns are sending. 

Might be related?

**Comment by void-gh** [comment]
Any update? Any additional information that is needed to track down this issue?

Thanks.

**Comment by void-gh** [comment]
So digging around a bit, we seem to have the same issue with sending segment (broadcast) emails.

Looking in to the Mautic spool queue (./mautic/app/spool/default) there were a bunch of *.message files sitting there from various past failed attempts.

Cleared out the queue, and created a brand new broadcast using the "Skyline" template results in new .message files being generated but do not actually go anywhere. Checking the mautic log I see the following error:

`[2018-03-09 00:00:02] mautic.NOTICE: Swift_IoException: Unable to open file for reading [blob:https://FQDN/e1fd1ded-09e8-403b-9aae-7fc9e9f5a8ed] (uncaught exception) at /var/www/mautic/vendor/swiftmailer/swiftmailer/lib/classes/Swift/ByteStream/FileByteStream.php line 144 while running console command ``mautic:emails:send`` [] []
[2018-03-09 00:30:01] mautic.NOTICE: Swift_IoException: Unable to open file for reading [blob:https://FQDN/e1fd1ded-09e8-403b-9aae-7fc9e9f5a8ed] (uncaught exception) at /var/www/mautic/vendor/swiftmailer/swiftmailer/lib/classes/Swift/ByteStream/FileByteStream.php line 144 while running console command ``mautic:emails:send`` [] []`


The cron jobs are being run as the web server user (www-data), and all the file system permissions appear to be good.

**Comment by pcottrell** [comment]
My issue was that there was a limit on how many emails could be sent to an email address and that quota was being met that's why some people would not receive email on the list. I bumped up the total amount of emails an email address could receive and they all sent fine

**Comment by alfredosimoes** [comment]
I have a similar problem. 

Emails are marked as sent by i have /mautic/app/spool/default full of messages to be sent. Messages dont get delivered.

My email is defined as queued and using my host email account. I am not if the queue definitions are correct. If i sent an individual email it gets delivered but my campaign are not delivered.

I think my cron jobs are correct:

/usr/local/bin/php /home/ascmico/public_html/mautic/app/console mautic:queue:process --env=prod
/usr/local/bin/php /home/ascmico/public_html/mautic/app/console mautic:messages:send --env=prod
/usr/local/bin/php /home/ascmico/public_html/mautic/app/console mautic:email:fetch --env=prod
/usr/local/bin/php /home/ascmico/public_html/mautic/app/console mautic:broadcasts:send --env=prod

There anyone that can help on this problem? Anyone knows how to bypass this situation?

Thanks in advance. Best regards.

Alfredo

**Comment by escopecz** [comment]
@alfredosimoes you are missing the one sending emails from the queue: https://mautic.org/docs/en/setup/cron_jobs.html#process-email-queue

**Comment by alfredosimoes** [comment]
Hi Jonh (escopecz) ! Thanks for your reply.
I am addind it.
/usr/local/bin/php /home/ascmico/public_html/mautic/app/console mautic:emails:send --env=prod
I will let you know the results asap.

**Comment by alfredosimoes** [comment]
This cron job never runs. It has the same path like the others that are running.
/usr/local/bin/php /home/ascmico/public_html/mautic/app/console mautic:emails:send --env=prod
Is it properly coded?

**Comment by escopecz** [comment]
Many people use it. What's the output when you execute that command manually?

**Comment by Bigood** [comment]
Exact same case here, 2.13.2-dev. Updating to 2.14.0-beta, will post an update soon.

**Comment by nisimjoseph** [comment]
I didn't experience an issue in 2.13.2. the issue started only when upgraded to 2.14.0.

**Comment by mleader1** [comment]
I regret I upgraded our mautic to the latest version - spent whole day trying to figure out why campaign was triggered successfully but no emails sent without a clue.

What I spotted was that any new subscribers created after the upgrade seems working ok. i.e. Mautic sends emails to those new subscribers for each campaign without a problem!!!  All existing contacts of the campaigns do not receive any email!!!

waiting for a solution desperately....

**Comment by nisimjoseph** [comment]
even new subscribers are not getting the emails. nothing sent since the update. I don't know what to do. 

Please HELP!
All Crons are ok and it just stops sending emails.

**Comment by mleader1** [comment]
@alanhartless   checked  .  There were a few that's worth noticing:

[2018-08-01 21:21:23] mautic.CRITICAL: Uncaught PHP Exception Doctrine\ORM\ORMInvalidArgumentException: "A new entity was found through the relationship 'MauticPlugin\MauticFocusBundle\Entity\Stat#lead' that was not configured to cascade persist operations for entity: Mautic\LeadBundle\Entity\Lead with ID #. To solve this issue: Either explicitly call EntityManager#persist() on this unknown entity or configure cascade persist  this association in the mapping for example @ManyToOne(..,cascade={"persist"})." at /--PATHREMOVED---/public_html/vendor/doctrine/orm/lib/Doctrine/ORM/ORMInvalidArgumentException.php line 92 {"exception":"[object] (Doctrine\\ORM\\ORMInvalidArgumentException(code: 0): A new entity was found through the relationship 'MauticPlugin\\MauticFocusBundle\\Entity\\Stat#lead' that was not configured to cascade persist operations for entity: Mautic\\LeadBundle\\Entity\\Lead with ID #. To solve this issue: Either explicitly call EntityManager#persist() on this unknown entity or configure cascade persist  this association in the mapping for example @ManyToOne(..,cascade={\"persist\"}). at /--PATHREMOVED--/vendor/doctrine/orm/lib/Doctrine/ORM/ORMInvalidArgumentException.php:92)"} []



There were also lots of notice errors but I guess is less critical.

I confirm the following:

1) all new subscribers created directly from the mautic back-end receives all emails triggered by the campaigns, however new subscribers imported via csv file as well as ALL existing subscribers before the upgrade they do NOT receive any emails from campaigns.    

2) Emails sent manually works so definitely not something to do with email providers

3) Campaign details page under "Actions" tab, it showed XXX number of completed actions for "Send or schedule email" events with small number of "Pending actions". However going into Email listing page I can see the "Stats" column it showed only extremely limited number of emails sent.  (we have thousands of valid contacts but only less than a hundred of emails sent - which are new contacts generated on daily basis).

4) Also checked /var/log/httpd  errors, all clean.

5) Also checked php.ini and my.cfg incase of server resource related issues - even putting it to max didn't solve this issue. (Note we have less than 5000 contacts so I guess we are not server resource hungry?)

6) Lastly - Also tried to backup db and restore on new environment - didn't work. So guess it must be something related to SQL queries or logics in the PHP code.

I don't mind to set up everything again on new environment but there's no full csv export/import for all existing generated data (custom fields, campaigns, emails, stats). So I can't easily abandon what we've got in existing db.

PLEASE HELP.  IT'S DRIVING ME MAD

**Comment by yarokifor** [comment]
@mleader1 I feel your pain. I've been trying to pin point this issue for a couple of days. I can confirm that it definitely has to do something with imported contacts from CSV file. One difference I see between the contacts create by the CSV import and others is that the owner_id is NULL in mysql. Can someone else confirm this as well?

**Comment by yarokifor** [comment]
I am also running v2.14.0 on PHP 7.0.

**Comment by socmov** [comment]
I'm having the same issue after upgrading to 2.14

**Comment by YosuCadilla** [comment]
I can confirm this is not solved on 2.14.1(beta)

**Comment by aderbalbotelho** [comment]
php5 -d memory_limit = 3G  app / console mautic: campaigns: trigger --force --batch-limit = 5
Firing event for campaign 1
Triggering Events for Recent Recent Contacts
% events% total event (s) to be processed in batches of 5 contacts
A total of {0} 0 events were executed
A total of {0} 0 events were scheduled

**Comment by YosuCadilla** [comment]
Adding #6156 to the list.
In that particular Issue we might have found what the problem is...

Seems the problem or bug triggers when the the action "Modify contact's segments" is used on a campaign.

Confirmed, I built a new campaign that wasn't using the Modify contact's segments and all emails did go out.

Workaround being: Use a field instead for controlling if user clicked on the email instead of removing them from the segment.

**Comment by javjim-mautic** [comment]
Can you comment how to reproduce the bug? @YosuCadilla

**Comment by YosuCadilla** [comment]
Yes, @javjim-mautic, I want to test with 2.14.1 (non beta) before opening yet another issue.
In the meantime we're trying to narrow it down on #6156.

When I have a clearer notion I will open a new issue and document it as much as possible.

**Comment by npracht** [comment]
i think we can consider the fixes #6482 #6211 & #6531 covers this issue. We'll reopen if necessary.
